core/src/main/java/org/elasticsearch/cluster/AbstractNamedDiffable.java

```diff
@@ -30,43 +31,48 @@
  * Abstract diffable object with simple diffs implementation that sends the entire object if object has changed or
  * nothing is object remained the same. Comparing to AbstractDiffable, this class also works with NamedWriteables
  */
-public abstract class AbstractNamedDiffable<T extends Diffable<T> & NamedWriteable> implements Diffable<T>, NamedWriteable {
+public abstract class AbstractNamedDiffable<T extends NamedDiffable<T>> implements Diffable<T>, NamedWriteable {
 
     @Override
     public Diff<T> diff(T previousState) {
         if (this.get().equals(previousState)) {
-            return new CompleteNamedDiff<>(previousState.getWriteableName());
+            return new CompleteNamedDiff<>(previousState.getWriteableName(), previousState.getMinimalSupportedVersion());
         } else {
             return new CompleteNamedDiff<>(get());
         }
     }
 
-    public static <T extends Diffable<T> & NamedWriteable> NamedDiff<T> readDiffFrom(Class<? extends T> tClass, String name, StreamInput in)
+    public static <T extends NamedDiffable<T>> NamedDiff<T> readDiffFrom(Class<? extends T> tClass, String name, StreamInput in)
         throws IOException {
         return new CompleteNamedDiff<>(tClass, name, in);
     }
 
-    private static class CompleteNamedDiff<T extends Diffable<T> & NamedWriteable> implements NamedDiff<T> {
+    private static class CompleteNamedDiff<T extends NamedDiffable<T>> implements NamedDiff<T> {
 
         @Nullable
         private final T part;
 
         private final String name;
 
+        @Nullable
+        private final Version minimalSupportedVersion;
```

Comment:
I think this deserves javadoc, especially because you expect it to be `null` sometimes. It'd be nice to list "sometimes". It looks like it really only needs to be non-null if you are going to write it but you make a best effort to read it if possible. Maybe it isn't worth doing that and just leaving it null if you read it.